### PR TITLE
Optimize u>=1 to u!=0 and u<1 to u==0 

### PR DESCRIPTION
**Pull request by EgorBo**
JIT already applies these tricks for signed types. This PR adds support for unsigned (extends existing logic). The diff looks big because I removed the surrounding if-else block
```csharp
using System.Runtime.CompilerServices;

class MyTests
{
    // make sure my changes don't regress signed types:
    bool C1(int a) => a >= 1;   // to "a > 0"
    bool C2(int a) => a < 1;    // to "a <= 0"
    bool C3(int a) => a <= -1;  // to "a < 0"
    bool C4(int a) => a > -1;   // to "a >= 0"

    // new cases:
    bool UC1(uint a) => a >= 1; // to "a != 0"
    bool UC2(uint a) => a < 1;  // to "a == 0"
}
```

```diff
; Method MyTests:C1(int):bool:this
G_M54558_IG02:
       test     edx, edx
       setg     al
       movzx    rax, al
; Total bytes of code: 9


; Method MyTests:C2(int):bool:this
G_M54557_IG02:
       test     edx, edx
       setle    al
       movzx    rax, al
; Total bytes of code: 9


; Method MyTests:C3(int):bool:this
G_M54556_IG02:
       test     edx, edx
       setl     al                  ; LLVM generates shr eax, 31
       movzx    rax, al
; Total bytes of code: 9


; Method MyTests:C4(int):bool:this
G_M54555_IG02:
       test     edx, edx
       setge    al
       movzx    rax, al
; Total bytes of code: 9


; Method MyTests:UC1(int):bool:this
G_M19883_IG02:
-      cmp      edx, 1
-      setae    al
+      test     edx, edx
+      setne    al
       movzx    rax, al
-; Total bytes of code: 10
+; Total bytes of code: 9


; Method MyTests:UC2(int):bool:this
G_M19880_IG02:
-      cmp      edx, 1
-      setb     al
+      test     edx, edx
+      sete     al
       movzx    rax, al
-; Total bytes of code: 10
+; Total bytes of code: 9
```
/cc: @mikedn

### CODE REVIEW

File: src/jit/morph.cpp [hunk 1]
```diff
@@ -12767,69 +12767,80 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
         case GT_GE:
         case GT_GT:
 
-            if ((tree->gtFlags & GTF_UNSIGNED) == 0)
+            if (op2->gtOper == GT_CNS_INT)
             {
-                if (op2->gtOper == GT_CNS_INT)
+                bool isOpSigned = (tree->gtFlags & GTF_UNSIGNED) == 0;
```
Comment (mikedn) [inline src/jit/morph.cpp]: use `tree->IsUnsigned()`

Comment (mikedn) [reply src/jit/morph.cpp]: And remove the `isOpSigned` variable. There's no point in doing CSE for the native C++ compiler...

File: src/jit/morph.cpp [hunk 2]
```diff
@@ -12767,69 +12767,80 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
...
                 {
-                    cns2 = op2;
-                    /* Check for "expr relop 1" */
-                    if (cns2->IsIntegralConst(1))
+                    /* Check for "expr >= 1" */
+                    if (oper == GT_GE)
                     {
-                        /* Check for "expr >= 1" */
-                        if (oper == GT_GE)
+                        if (isOpSigned)
```
Comment (mikedn) [inline src/jit/morph.cpp]: Not a fan of ternary operator in general but in this case it looks just like what the doctor ordered.

File: src/jit/morph.cpp [hunk 3]
```diff
@@ -12767,69 +12767,80 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
...
-                            // IF we get here we should be changing 'oper'
-                            assert(tree->OperGet() != oper);
+                            /* Change to "expr == 0" */
+                            oper = GT_EQ;
+                        }
+                        goto SET_OPER;
+                    }
+                }
+                /* Check for "expr relop -1" */
+                if (isOpSigned && cns2->IsIntegralConst(-1) && ((oper == GT_LE) || (oper == GT_GT)))
```
Comment (mikedn) [inline src/jit/morph.cpp]: Is it me or this needs an `else`? If it was 1 it can't also be -1.

File: src/jit/morph.cpp [hunk 4]
```diff
@@ -12767,69 +12767,80 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
...
                         }
+
+                        op2 = tree->gtOp.gtOp2 = gtFoldExpr(op2);
                     }
                 }
-            }
-            else // we have an unsigned comparison
-            {
-                if (op2->IsIntegralConst(0))
+                if (!isOpSigned && op2->IsIntegralConst(0))
```
Comment (mikedn) [inline src/jit/morph.cpp]: else ?

**Comment by mikedn** [comment]
Any diffs?

**Comment by EgorBo** [comment]
@mikedn is there a doc on how to do it properly for the whole coreclr/corefx? I am only using `COMPlus_JitDisasm`. And my own add-in for VS for simple diffs 🙂:

fgege

**Comment by mikedn** [comment]
For diffs you need jit-diff from the jitutils repository: https://github.com/dotnet/jitutils/blob/master/doc/diffs.md

Running jit-diff itself shouldn't be too difficult but you need a coreclr "baseline". The docs recommend using a separate coreclr clone but I find that to be overkill. I normally just build master and save the clrjit.dll in a "base" folder and pass that to jit-diff.

**Comment by mikedn** [comment]
Here's the x64 FX diff:
```
Found 3 files with textual diffs.
Crossgen Diffs for System.Private.CoreLib.dll, framework assemblies for x64 default jit
Summary:
(Lower is better)
Total bytes of diff: -280 (-0.00% of base)
    diff is an improvement.
Top file improvements by size (bytes):
        -272 : System.Private.CoreLib.dasm (-0.01% of base)
          -8 : System.Private.Xml.dasm (-0.00% of base)
2 total files with size differences (2 improved, 0 regressed), 127 unchanged.
Top method improvements by size (bytes):
        -198 (-4.88% of base) : System.Private.CoreLib.dasm - Vector64`1:ToString():ref:this (11 methods)
         -52 (-3.44% of base) : System.Private.CoreLib.dasm - Vector64`1:Equals(struct):bool:this (11 methods)
         -10 (-11.36% of base) : System.Private.CoreLib.dasm - Vector64:GetElement(struct,int):long (2 methods)
          -7 (-0.92% of base) : System.Private.Xml.dasm - BigNumber:Mul(byref):this
          -6 (-0.61% of base) : System.Private.CoreLib.dasm - Vector64`1:GetHashCode():int:this (11 methods)
Top method improvements by size (percentage):
         -10 (-11.36% of base) : System.Private.CoreLib.dasm - Vector64:GetElement(struct,int):long (2 methods)
          -5 (-11.11% of base) : System.Private.CoreLib.dasm - Vector64:GetElement(struct,int):double
        -198 (-4.88% of base) : System.Private.CoreLib.dasm - Vector64`1:ToString():ref:this (11 methods)
         -52 (-3.44% of base) : System.Private.CoreLib.dasm - Vector64`1:Equals(struct):bool:this (11 methods)
          -7 (-0.92% of base) : System.Private.Xml.dasm - BigNumber:Mul(byref):this
8 total methods with size differences (8 improved, 0 regressed), 146770 unchanged.
1 files had text diffs but not size diffs.
Microsoft.CodeAnalysis.VisualBasic.dasm had 12 diffs
```

File: src/jit/morph.cpp [hunk 5]
```diff
@@ -12767,69 +12767,63 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac)
...
+                    else if (oper == GT_GT)
+                    {
+                        /* Change to "expr >= 0" */
+                        oper = GT_GE;
 
-                            // Keep the old ValueNumber for 'tree' as the new expr
-                            // will still compute the same value as before
-                            tree->SetOper(oper, GenTree::PRESERVE_VN);
-                            cns2->gtIntCon.gtIconVal = 0;
+                    SET_OPER:
```
Comment (ArtBlnd) [inline src/jit/morph.cpp]: Maybe better replace it to outside of `else if (!tree->IsUnsigned() && cns2->IsIntegralConst(-1))` and also its end of `if (op2->gtOper == GT_CNS_INT)`.

**Comment by BruceForstall** [comment]
@dotnet/jit-contrib

**Comment by EgorBo** [comment]
@sandreenko sure!

**Comment by sandreenko** [comment]
@EgorBo could you please clone https://github.com/dotnet/jitutils, run bootstrap.cmd and then run jit-format to fix the formatting errors, like:
```
seandree@SEANDREEPC C:\git\tools\jitutils\bin
$ jit-format.bat --fix -c E:\git\coreclr
```

**Pull request merged**